pytorch_lightning/trainer/supporters.py

```diff
@@ -119,27 +122,249 @@ def mean(self):
         return self.total / self.num_values
 
 
+class LightningBatchSamplerWrapper:
+    """
+    This class wraps user batch sampler, so we can extract
+    the batch_indices for tracking each sample.
+    """
+
+    def __init__(self, batch_sampler):
+        self.batch_sampler = batch_sampler
+        self.batch_indices = None
+
+    def __iter__(self):
+        for batch_indices in self.batch_sampler:
+            self.batch_indices = batch_indices
+            yield batch_indices
+
+    @staticmethod
+    def to_new_dataloader(dataloader: DataLoader) -> DataLoader:
+        """
+        This function will wrap the user batch_sampler to track the returned batch indices
+        """
+        if not isinstance(dataloader, DataLoader):
+            raise MisconfigurationException('Autoid only works with torch dataloaders or derived classes!')
+
+        if isinstance(dataloader.dataset, IterableDataset):
+            return dataloader
+
+        skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
+        skip_valid_keys = ['args', 'kwargs', 'self']
+
+        params = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
+        contains_dataset = True
+
+        if type(dataloader) is not DataLoader:
+            contains_dataset = "dataset" in valid_kwargs
+            valid_kwargs.update(inspect.signature(DataLoader.__init__).parameters)
+
+        dl_args = {
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in skip_keys
+        }
+
+        multiprocessing_context = dataloader.multiprocessing_context
+
+        # override parameters to enable batch_sampler injection
+        dl_args["batch_size"] = 1
+        dl_args["sampler"] = None
+        dl_args["shuffle"] = None
+        dl_args["batch_sampler"] = LightningBatchSamplerWrapper(dataloader.batch_sampler)
+        dl_args["drop_last"] = False
+        dl_args['multiprocessing_context'] = multiprocessing_context
```

Comment:
```suggestion
        dl_args.update({
            "batch_size": 1,
            "sampler": None,
            "shuffle": None,
            "batch_sampler": LightningBatchSamplerWrapper(dataloader.batch_sampler),
            "drop_last": False
            "multiprocessing_context": multiprocessing_context,
        })
```
easier to see that you update still the same variable